File: bench/vec3.c
/* C twin of vec3.ts: the same methods as static functions over a heap struct. */
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#define N 50000000 /* bench:n */

typedef struct {
  double x, y, z;
} Vec3;

static Vec3 *vec3(double x, double y, double z) {
  Vec3 *v = malloc(sizeof *v);
  v->x = x; v->y = y; v->z = z;
  return v;
}
static void add(Vec3 *t, const Vec3 *o) {
  t->x = t->x + o->x;
  t->y = t->y + o->y;
  t->z = t->z + o->z;
}
static void addScaled(Vec3 *t, const Vec3 *o, double s) {
  t->x = t->x + o->x * s;
  t->y = t->y + o->y * s;
  t->z = t->z + o->z * s;
}
static double dot(const Vec3 *t, const Vec3 *o) { return t->x * o->x + t->y * o->y + t->z * o->z; }
static void crossInto(const Vec3 *t, const Vec3 *o, Vec3 *out) {
  out->x = t->y * o->z - t->z * o->y;
  out->y = t->z * o->x - t->x * o->z;
  out->z = t->x * o->y - t->y * o->x;
}
static double norm(const Vec3 *t) { return sqrt(dot(t, t)); }

int main(void) {
  const double DT = 0.0000001;
  Vec3 *p = vec3(0, 0, 0);
  Vec3 *v = vec3(1, 2, 3);
  Vec3 *g = vec3(0, -0.0000001, 0);
  Vec3 *kick = vec3(0, 0, 0);
  double energy = 0;
  for (int32_t i = 0; i < N; i++) {
    add(v, g);
    crossInto(v, g, kick);
    addScaled(v, kick, 0.001);
    addScaled(p, v, DT);
    energy = energy + 0.5 * dot(v, v) + norm(p) * DT;
  }
  printf("%.17g\n%.17g\n%.17g\n%.17g\n", p->x, p->y, p->z, energy);
  return 0;
}

File: bench/vec3.ts
// Struct-heavy loop (f64 mode): a Vec3 class with methods, called N times in
// a tight loop that integrates a point under a constant force with a small
// perpendicular kick. Every method mutates in place or returns a number, so
// the loop allocates nothing (the arena would otherwise grow by 24 bytes per
// temporary; escape-analysed stack allocation of such temporaries is WP6).
// Prints the final position and the accumulated "energy"; all languages use
// the same expression order so the doubles agree bit for bit.
// Compile with --number-mode f64.
class Vec3 {
  x: number;
  y: number;
  z: number;
  constructor(x: number, y: number, z: number) {
    this.x = x;
    this.y = y;
    this.z = z;
  }
  add(o: Vec3): void {
    this.x = this.x + o.x;
    this.y = this.y + o.y;
    this.z = this.z + o.z;
  }
  addScaled(o: Vec3, s: number): void {
    this.x = this.x + o.x * s;
    this.y = this.y + o.y * s;
    this.z = this.z + o.z * s;
  }
  scale(s: number): void {
    this.x = this.x * s;
    this.y = this.y * s;
    this.z = this.z * s;
  }
  dot(o: Vec3): number {
    return this.x * o.x + this.y * o.y + this.z * o.z;
  }
  crossInto(o: Vec3, out: Vec3): void {
    out.x = this.y * o.z - this.z * o.y;
    out.y = this.z * o.x - this.x * o.z;
    out.z = this.x * o.y - this.y * o.x;
  }
  norm(): number {
    return Math.sqrt(this.dot(this));
  }
}

export function main(): i32 {
  const N: i32 = 50000000; // bench:n
  const DT = 0.0000001;
  const p = new Vec3(0, 0, 0);
  const v = new Vec3(1, 2, 3);
  const g = new Vec3(0, -0.0000001, 0);
  const kick = new Vec3(0, 0, 0);
  let energy = 0;
  for (let i: i32 = 0; i < N; i++) {
    v.add(g);
    v.crossInto(g, kick);
    v.addScaled(kick, 0.001);
    p.addScaled(v, DT);
    energy = energy + 0.5 * v.dot(v) + p.norm() * DT;
  }
  console.log(p.x);
  console.log(p.y);
  console.log(p.z);
  console.log(energy);
  return 0;
}
